## Pandas Challenge - Heroes of Pymoli

In [92]:
# Dependencies and Setup
import pandas as pd
import numpy as np

# Create the path to the CSV file
pymoli_path = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(pymoli_path)
purchase_data.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

### Player Count

* Display the total number of players

In [93]:
# Calculate the total number of unique players in the SN column and print result
total_players = len(purchase_data["SN"].unique())
print("Total Players =",total_players)

Total Players = 576


### Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.
* Create a summary data frame to hold the results
* Optional: give the displayed data cleaner formatting
* Display the summary data frame

In [94]:
# Return a count of the number of unique items in the Item ID column
unique_items = len(purchase_data["Item ID"].unique())

# Find the average price of purchased items and format to give $ value to 2 decimal places.
average_price = purchase_data["Price"].sum()/purchase_data["Price"].count()
formatted_average_price = "${:,.2f}".format(average_price)

# Count the number of purchases.
number_of_purchases = len(purchase_data["Purchase ID"])

# Find the total revenue by summing the Price column and format to give $ value to 2 decimal places.
total_revenue = purchase_data["Price"].sum()
formatted_total_revenue = "${:,.2f}".format(total_revenue)

# Create a new DataFrame of the above results and display the summary table.
summary_df=pd.DataFrame({"Number of Unique Items":unique_items,"Average Price":formatted_average_price,
                         "Number of Purchases":number_of_purchases,"Total Revenue":formatted_total_revenue}, index=[0])
summary_df

Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     179         $3.05                  780     $2,379.77

### Gender Demographics

* Percentage and Count of Male Players
* Percentage and Count of Female Players
* Percentage and Count of Other / Non-Disclosed

In [110]:
# Group the raw data by gender
grouped_df=purchase_data.groupby("Gender")

# Count each gender using nunique (to return a scalar value) to avoid double up of SN's
count_gender = grouped_df.nunique()["SN"]

# Calculate the gender percentage using count_gender and total_players (found in an earlier step)
percentage_gender = count_gender / total_players * 100

# Create a new DataFrame to store results
gender_summary = pd.DataFrame({"Total Count": count_gender,"Percentage of Players": percentage_gender})

# Format the Percentage Players column to give a % value with 2 decimal places
gender_summary["Percentage of Players"]=gender_summary["Percentage of Players"].map("{0:.2f}%".format)

# Display summary table
gender_summary

Total Count Percentage of Players
Gender                                                  
Female                          81                14.06%
Male                           484                84.03%
Other / Non-Disclosed           11                 1.91%


### Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender
* Create a summary data frame to hold the results
* Optional: give the displayed data cleaner formatting
* Display the summary data frame

In [111]:
# Re-use the previous groupby by Gender (grouped_df) for this section and calculate the count of purchases.
purchase_count = grouped_df["Purchase ID"].count()

# Calculate the average purchase price.
ave_purchase_price = grouped_df["Price"].mean()

# Calculate the total purchase price.
total_purchase_price = grouped_df["Price"].sum()

# Calculate the Avg Total Purchase per person
total_purchase_per_person=total_purchase_price/grouped_df.nunique()["SN"]

# Create a new DataFrame with the above results
purchasing_summary = pd.DataFrame({"Purchase Count": purchase_count,"Average Purchase Price": ave_purchase_price,
                                  "Total Purchase Value":total_purchase_price, "Avg Total Purchase per Person":total_purchase_per_person})

# Format all three $ columns to give a $ value with 2 decimal places
purchasing_summary["Average Purchase Price"]=purchasing_summary["Average Purchase Price"].map("${0:,.2f}".format)
purchasing_summary["Total Purchase Value"]=purchasing_summary["Total Purchase Value"].map("${0:,.2f}".format)
purchasing_summary["Avg Total Purchase per Person"]=purchasing_summary["Avg Total Purchase per Person"].map("${0:,.2f}".format)

# Display summary table
purchasing_summary

Purchase Count Average Purchase Price  \
Gender                                                         
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Avg Total Purchase per Person  
Gender                                                                    
Female                             $361.94                         $4.47  
Male                             $1,967.64                         $4.07  
Other / Non-Disclosed               $50.19                         $4.56

### Age Demographics

* Establish bins for ages
* Categorize the existing players using the age bins. Hint: use pd.cut()
* Calculate the numbers and percentages by age group
* Create a summary data frame to hold the results
* Optional: round the percentage column to two decimal points
* Display Age Demographics Table

In [117]:
# Create the bins in which the data will be held   
bins = [0, 9.9, 14.9, 19.9, 24.9, 29.9, 34.9, 39.9, 999]

# Create the names for the bins
group_names = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34","35-39","40+"]

# Segment the data and add new column for age group
purchase_data["Age Group"] = pd.cut(purchase_data["Age"], bins, labels=group_names, include_lowest=False)
purchase_data.head()

# Group the raw data by age group
age_grouped_df=purchase_data.groupby("Age Group")

# Count the number of players in each age group using nunique (to return a scalar value) to avoid double up of SNs
count_age = age_grouped_df.nunique()["SN"]

# Calculate the age group percentage using count_age and total_players
percentage_age = count_age / total_players * 100

# Create a new DataFrame for the results
age_summary = pd.DataFrame({"Total Count": count_age,"Percentage of Players": percentage_age})

# Format the Percentage Players column to give a % value with 2 decimal places
age_summary["Percentage of Players"]=age_summary["Percentage of Players"].map("{0:.2f}%".format)

# Display summary table
age_summary

Total Count Percentage of Players
Age Group                                   
<10                 17                 2.95%
10-14               22                 3.82%
15-19              107                18.58%
20-24              258                44.79%
25-29               77                13.37%
30-34               52                 9.03%
35-39               31                 5.38%
40+                 12                 2.08%

### Purchasing Analysis (Age)

* Bin the purchase_data data frame by age
* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below
* Create a summary data frame to hold the results
* Optional: give the displayed data cleaner formatting
* Display the summary data frame

In [118]:
# Re-use the previous groupby by age group (age_grouped_df) for this section and calculate the count of purchases.
age_purchase_count = age_grouped_df["Purchase ID"].count()

# Calculate the average purchase price.
age_ave_purchase_price = age_grouped_df["Price"].mean()

# Calculate the total purchase price.
age_total_purchase_price = age_grouped_df["Price"].sum()

# Calculate the Avg Total Purchase per person
age_total_purchase_per_person=age_total_purchase_price/age_grouped_df.nunique()["SN"]

# Create a new DataFrame with the above results
age_purchasing_summary = pd.DataFrame({"Purchase Count": age_purchase_count,"Average Purchase Price": age_ave_purchase_price,
                                  "Total Purchase Value":age_total_purchase_price, "Avg Total Purchase per Person":age_total_purchase_per_person})

# Format all three $ columns to give a $ value with 2 decimal places
age_purchasing_summary["Average Purchase Price"]=age_purchasing_summary["Average Purchase Price"].map("${0:,.2f}".format)
age_purchasing_summary["Total Purchase Value"]=age_purchasing_summary["Total Purchase Value"].map("${0:,.2f}".format)
age_purchasing_summary["Avg Total Purchase per Person"]=age_purchasing_summary["Avg Total Purchase per Person"].map("${0:,.2f}".format)

# Display summary table
age_purchasing_summary

Purchase Count Average Purchase Price Total Purchase Value  \
Age Group                                                               
<10                    23                  $3.35               $77.13   
10-14                  28                  $2.96               $82.78   
15-19                 136                  $3.04              $412.89   
20-24                 365                  $3.05            $1,114.06   
25-29                 101                  $2.90              $293.00   
30-34                  73                  $2.93              $214.00   
35-39                  41                  $3.60              $147.67   
40+                    13                  $2.94               $38.24   

          Avg Total Purchase per Person  
Age Group                                
<10                               $4.54  
10-14                             $3.76  
15-19                             $3.86  
20-24                             $4.32  
25-29                             $3.81  
30-34                             $4.12  
35-39                             $4.76  
40+                               $3.19

### Top Spenders

* Run basic calculations to obtain the results in the table below
* Create a summary data frame to hold the results
* Sort the total purchase value column in descending order
* Optional: give the displayed data cleaner formatting
* Display a preview of the summary data frame

In [123]:
# Group data by SN
SN_grouped_df=purchase_data.groupby("SN")

# Calculate purchase count per SN
SN_purchase_count = SN_grouped_df["Purchase ID"].count()

# Calculate average purchase price per SN
SN_average_price = SN_grouped_df["Price"].mean()

# Calculate total purchase value per SN
SN_total_price = SN_grouped_df["Price"].sum()

# Create new DataFrame for above data
top_spender_df = pd.DataFrame({"Purchase Count":SN_purchase_count,"Average Purchase Price":SN_average_price,
                              "Total Purchase Value":SN_total_price})

# Sort the Total Purchase Value column in descending order
top_spender_sorted_df = top_spender_df.sort_values("Total Purchase Value",ascending=False)

# Format the two price columns
top_spender_sorted_df["Average Purchase Price"] = top_spender_sorted_df["Average Purchase Price"].map("${0:.2f}".format)
top_spender_sorted_df["Total Purchase Value"] = top_spender_sorted_df["Total Purchase Value"].map("${0:.2f}".format)

# Display the top 5 spenders
top_spender_sorted_df.head(5)

Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

### Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns
* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value
* Create a summary data frame to hold the results
* Sort the purchase count column in descending order
* Optional: give the displayed data cleaner formatting
* Display a preview of the summary data frame

In [130]:
# Retrieve the Item ID, Item Name, and Price columns
items_df = purchase_data[["Item ID","Item Name","Price"]]

# Groupby Item ID and Item Name. Perform count on purchased items, item price and total purchase value.
grouped_items_df = items_df.groupby(["Item ID","Item Name"])
item_count = grouped_items_df["Item ID"].count()
item_price = grouped_items_df["Price"].mean()
total_value = grouped_items_df["Price"].sum()

# Create new DataFrame to store the above data
item_summary_df=pd.DataFrame({"Purchase Count":item_count,"Item Price":item_price,"Total Purchase Value":total_value})

# Sort the purchase count column in descending order
item_summary_sorted_df = item_summary_df.sort_values("Purchase Count",ascending=False)

# Format the two price columns
item_summary_sorted_df["Item Price"] = item_summary_sorted_df["Item Price"].map("${0:.2f}".format)
item_summary_sorted_df["Total Purchase Value"] = item_summary_sorted_df["Total Purchase Value"].map("${0:.2f}".format)

# Display the top 5 most purchased items
item_summary_sorted_df.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
132     Persuasion                                                 9   
108     Extraction, Quickblade Of Trembling Hands                  9   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
132     Persuasion                                        $3.22   
108     Extraction, Quickblade Of Trembling Hands         $3.53   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
132     Persuasion                                                 $28.99  
108     Extraction, Quickblade Of Trembling Hands                  $31.77

### Most Profitable Items

* Sort the above table by total purchase value in descending order
* Optional: give the displayed data cleaner formatting
* Display a preview of the data frame

In [134]:
# Sort the purchase count column in descending order
item_summary_sorted_again_df = item_summary_df.sort_values("Total Purchase Value",ascending=False)

# Format the two price columns
item_summary_sorted_again_df["Item Price"] = item_summary_sorted_again_df["Item Price"].map("${0:.2f}".format)
item_summary_sorted_again_df["Total Purchase Value"] = item_summary_sorted_again_df["Total Purchase Value"].map("${0:.2f}".format)

# Display top 5 most profitable items
item_summary_sorted_again_df.head(5)

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
103     Singed Scalpel                                             8   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
103     Singed Scalpel                                             $34.80

## Conclusions

The gender demographic that plays Heroes of Pymoli is overwhelmingly male at 84% of players, with only 14% female and <2% other/non-disclosed. However, male players spend the least on in-game purchases with females and other/non-disclosed players spending ~12.5% more on average per player than their male counterparts. 

Players in the 20-24 year age group make up the majority of the market at 45.8%, followed by 15-19 years old’s at 18.6% and 25-29 year old’s at 13.4%. So, these 3 age groups make up 76.7% of the players. Interestingly, the biggest spenders on in-game purchases are 35-39 year old’s and <10 year old’s at $4.76 and $4.54 average total purchase per person, respectively.

Based on the above, targeted marketing towards families, particularly towards mothers bonding with their kids over Heroes of Pymoli to get through the day may be a good strategy to increase profits from in-game purchases.

To improve value in the analysis of item purchases (most popular/profitable), a recommendation would be to add an 'item class'. This would provide information on whether a particular item class is where players find the most value to get ahead in the game i.e., particular weapon type, armour, tools, potions etc. This would provide valuable insight to the game developers on where the item demand is for future item additions.  